### ISSUE TITLE
Liveplot explodes in hutch-python on NFS

### ISSUE THREAD

**Issue by ZLLentz** [pcdshub/hutch-python#210]
When doing a bluesky scan in a hutch python session, after initializing the qt teleporter, something bad happens that causes a segfault. The errors suggest that the qt teleporter isn't doing the right thing.

Everything works great in the same environment if:
- you load a normal (not hutch-python) ipython session
- you do it on my desktop machine (AFS)

Can reproduce by:
- ssh psbuild-rhel7
- source pcds_conda
- hutch-python
- set up the qt teleporter (should be done inside hutch python but it is new enough that it isn't yet)
- try to do any scan that would produce a liveplot from bec

**Comment by aegger13** [comment]
tentatively assigning to myself, but will likely ask for help

**Comment by ZLLentz** [comment]
Minimal example script that breaks:
```
In [1]: from ophyd.sim import motor1, det1

In [2]: RE(bp.scan([det1], motor1, -10, 10, 21))
```

**Comment by ZLLentz** [comment]
There is also a new `initialize_qt_teleporter` function that needs to be called before anything works:
```
from bluesky.callbacks.mpl_plotting import initialize_qt_teleporter

initialize_qt_teleporter()
```

**Comment by ZLLentz** [comment]
With the above, everything works in a fresh ipython session but fails in a hutch python session.

**Comment by klauer** [comment]
May be relevant:
https://github.com/pcdshub/hutch-python/blob/d7920dcb62be21b77c477d442fa79606ba155dff/hutch_python/cli.py#L117-L118

**Comment by ZLLentz** [comment]
That's how you configure ipython to load with the qt backend without requiring the user to pass %matplotlib qt5. I agree that it is one possibility for the source of the issue.

Other things to check:
- Does the timing of the qt teleporter matter? E.g. does it have to be done before BEC?
- Does this happen when using a fresh RE/BEC rather than the ones hutch-python gives you?

**Comment by aegger13** [comment]
Update
I spent some time digging into this.  As a first pass, it looks like there is an issue with tk internals for BestEffortCallback not being called on main thread.  Why this happens only on NFS seems very strange.  
Teleporter doesn't seem to have any impact whatsoever. 
 I wanted to try and use 'Agg' for Matplotlib backend as it seems to deal better in threaded environments, but I get complaints about it already being set to 'Qt5Agg'.  
This is similar to a bug that was already fixed for bec in Bluesky https://github.com/bluesky/bluesky/pull/1232/files, although the fix doesn't fill me with confidence that it is well understood.

If anyone has an idea what this behavior would be different on NFS machines that would be a huge help.

**Comment by aegger13** [comment]
Here is a good stack trace of what's going on
Attribute Qt::AA_EnableHighDpiScaling must be set before QCoreApplication is created.
QObject::setParent: Cannot set parent, new parent is in a different thread
QObject::setParent: Cannot set parent, new parent is in a different thread
ERROR    An exception raised in the callback <function BestEffortCallback.__call__ at 0x7ff338f48598> is being suppressed to not interrupt plan execution.  To investigate try setting the BLUESKY_DEBUG_CALLBACKS env to '1'
Traceback (most recent call last):
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/core.py", line 56, in inner
    return func(*args, **kwargs)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py", line 98, in __call__
    super().__call__(name, doc, *args, **kwargs)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/mpl_plotting.py", line 74, in __call__
    return CallbackBase.__call__(self, name, doc)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/event_model/__init__.py", line 76, in __call__
    return self._dispatch(name, doc, validate)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/event_model/__init__.py", line 84, in _dispatch
    output_doc = getattr(self, name)(doc)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py", line 274, in descriptor
    peak_results=self.peaks)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/best_effort.py", line 536, in __init__
    super().__init__(*args, **kwargs)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/mpl_plotting.py", line 119, in __init__
    super().__init__(use_teleporter=kwargs.pop('use_teleporter', None))
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/mpl_plotting.py", line 65, in __init__
    self.__teleporter = _get_teleporter()
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/bluesky/callbacks/mpl_plotting.py", line 42, in _get_teleporter
    "A bluesky QtAwareCallback was instantiated from a background "
RuntimeError: A bluesky QtAwareCallback was instantiated from a background thread before the bluesky qt 'teleporter' was created. To avoid this issue, call bluesky.callbacks.mpl_plotting.initialize_qt_teleporter() from the main thread first.

**Comment by aegger13** [comment]
Also, if we try to use matplotlib in ipy session after setting up bec
n [4]: %matplotlib qt5
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-64557277dbff> in <module>
----> 1 get_ipython().run_line_magic('matplotlib', 'qt5')

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/IPython/core/interactiveshell.py in run_line_magic(self, magic_name, line, _stack_depth)
   2324                 kwargs['local_ns'] = sys._getframe(stack_depth).f_locals
   2325             with self.builtin_trap:
-> 2326                 result = fn(*args, **kwargs)
   2327             return result
   2328

<decorator-gen-108> in matplotlib(self, line)

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    185     # but it's overkill for just that one bit of state.
    186     def magic_deco(arg):
--> 187         call = lambda f, *a, **k: f(*a, **k)
    188
    189         if callable(arg):

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/IPython/core/magics/pylab.py in matplotlib(self, line)
     97             print("Available matplotlib backends: %s" % backends_list)
     98         else:
---> 99             gui, backend = self.shell.enable_matplotlib(args.gui.lower() if isinstance(args.gui, str) else args.gui)
    100             self._show_matplotlib_backend(args.gui, backend)
    101

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/IPython/core/interactiveshell.py in enable_matplotlib(self, gui)
   3429                 gui, backend = pt.find_gui_and_backend(self.pylab_gui_select)
   3430
-> 3431         pt.activate_matplotlib(backend)
   3432         pt.configure_inline_support(self, backend)
   3433

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/IPython/core/pylabtools.py in activate_matplotlib(backend)
    320     from matplotlib import pyplot as plt
    321
--> 322     plt.switch_backend(backend)
    323
    324     plt.show._needmain = False

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/matplotlib/pyplot.py in switch_backend(newbackend)
    227     Calling this command will close all open windows.
    228     """
--> 229     close('all')
    230     global _backend_mod, new_figure_manager, draw_if_interactive, _show
    231     matplotlib.use(newbackend, warn=False, force=True)

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/matplotlib/pyplot.py in close(*args)
    645         arg = args[0]
    646         if arg == 'all':
--> 647             _pylab_helpers.Gcf.destroy_all()
    648         elif isinstance(arg, six.integer_types):
    649             _pylab_helpers.Gcf.destroy(arg)

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/matplotlib/_pylab_helpers.py in destroy_all(cls)
     76         for manager in list(cls.figs.values()):
     77             manager.canvas.mpl_disconnect(manager._cidgcf)
---> 78             manager.destroy()
     79
     80         cls._activeQue = []

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/site-packages/matplotlib/backends/_backend_tk.py in destroy(self, *args)
    557             #self.toolbar.destroy()
    558             if self.canvas._idle_callback:
--> 559                 self.canvas._tkcanvas.after_cancel(self.canvas._idle_callback)
    560             self.window.destroy()
    561         if Gcf.get_num_fig_managers()==0:

/reg/g/pcds/pyps/conda/py36/envs/pcds-3.3.4/lib/python3.6/tkinter/__init__.py in after_cancel(self, id)
    773                              'after or after_idle')
    774         try:
--> 775             data = self.tk.call('after', 'info', id)
    776             script = self.tk.splitlist(data)[0]
    777             self.deletecommand(script)

RuntimeError: main thread is not in main loop

**Comment by ZLLentz** [comment]
The teleporter did have an impact for me- there was a segfault with the teleporter instead of just a traceback. But the teleporter only works with qt, so I figure that bluesky is assuming qt for its callbacks here.

**Comment by aegger13** [comment]
I'll have to check that again.  It seemed like the traceback was intermittent (felt very race condition like), but it was probably tied to the teleporter.  I think I might need to do more introspection into BestEffortCallback in a local copy of blue sky to get a better idea of what's happening.

**Comment by aegger13** [comment]
Update.  I know how to fix this in blue sky, but the overall approach is less than desirable.  Long story short: we get a seg fault when we hit the line.  There is no way to pass use_teleporter to bec.

if use_teleporter is None:
            import matplotlib
            use_teleporter = 'qt' in matplotlib.get_backend().lower()

As I expected from seeing that issue on GitHub
The way to work around this is to actually have a way to pass use_teleporter inside of bec initialization as bec inherits from QtAwareCallback.  I believe this is a deeper bug in the way that initialize_qt_teleporter works, and the way someone hacked the QtAwareCallback initialization.  I have a fix available for blue sky and hutch-python, but I'm going to spend a bit of time to see if I can make that code a bit cleaner/well organized, because the current fix is too hacky.

**Comment by ZLLentz** [comment]
I'm confused, the code shown here looks like it works fine so long as we set the matplotlib backend earlier? Therefore, the easiest solution would be to just call `matlplotlib.use` prior to making our BEC?

**Comment by ZLLentz** [comment]
Closed by #215
